toolchain/lower/file_context.cpp

```diff
@@ -121,6 +137,219 @@ auto FileContext::Run() -> std::unique_ptr<llvm::Module> {
+        // that are dependent on the specific. This information has been stored
+        // in lowered_states while lowering each function body.
+        if (AreFunctionBodiesEquivalent(specifics[i], specifics[j],
+                                        visited_equivalent_specifics,
+                                        visited_equivalent_specifics_flipped)) {
+          Set<SemIR::SpecificId> specifics_to_delete;
+          visited_equivalent_specifics.ForEach(
+              [&](std::pair<SemIR::SpecificId, SemIR::SpecificId>
+                      equivalent_entry) {
+                CARBON_VLOG("Found equivalent specifics: {0}, {1}",
+                            equivalent_entry.first, equivalent_entry.second);
+                ProcessSpecificEquivalence(equivalent_entry,
+                                           specifics_to_delete);
+              });
+
+          specifics_to_delete.ForEach([&](SemIR::SpecificId specific_id) {
+            DeleteFunctionSpecific(specific_id);
+          });
+
+          // Removed the replaced specific from the list of emitted specifics.
+          // Only the top level, since the others are somewhere else in the
+          // vector, they will be found and removed during processing.
+          specifics[j] = specifics[specifics.size() - 1];
+          specifics.pop_back();
+          --j;
+        } else {
+          // Only mark non-equivalence based on state for starting specifics.
+          auto insert_result = non_equivalent_specifics_.Insert(pair);
+          // Save a lookup; if pair was previously inserted, so was the flipped
+          if (insert_result.is_inserted()) {
+            non_equivalent_specifics_.Insert(flipped_pair);
+          }
+        }
+      }
+    }
+  }
+}
+
+auto FileContext::ProcessSpecificEquivalence(
+    std::pair<SemIR::SpecificId, SemIR::SpecificId>& pair,
+    Set<SemIR::SpecificId>& specifics_to_delete) -> void {
+  auto [specific1, specific2] = pair;
+  CARBON_CHECK(specific1.has_value() && specific2.has_value(),
+               "Expected values in equivalence check");
+
+  SemIR::SpecificId first_id =
+      equivalent_specifics_[specific1.index].has_value()
+          ? equivalent_specifics_[specific1.index]
+          : specific1;
+  SemIR::SpecificId second_id =
+      equivalent_specifics_[specific2.index].has_value()
+          ? equivalent_specifics_[specific2.index]
+          : specific2;
+  auto replace_with = (first_id.index < second_id.index) ? first_id : second_id;
+  auto to_replace = (first_id.index >= second_id.index) ? first_id : second_id;
+  equivalent_specifics_[specific1.index] = replace_with;
+  equivalent_specifics_[specific2.index] = replace_with;
+
+  if (replace_with != to_replace) {
+    is_replaced_specific_[to_replace.index] = true;
+    specific_functions_[to_replace.index]->replaceAllUsesWith(
+        specific_functions_[replace_with.index]);
+    specifics_to_delete.Insert(to_replace);
+  }
```

Comment:
WDYT of something like:

```suggestion
  if (first_id == second_id) {
    // Already equivalent.
    // TODO: Can these cases be prevented from reaching this call?
    return;
  }
  if (first_id.index >= second_id.index) {
    // Prefer the earlier index for canonical values.
    std::swap(first_id, second_id);
  }
  
  // Mark the equivalence.
  equivalent_specifics_[second_id.index] = first_id;
  is_replaced_specific_[second_id.index] = true;
  specific_functions_[second_id.index]->replaceAllUsesWith(
      specific_functions_[first_id.index]);
  specifics_to_delete.Insert(second_id);
```